app/api/items/ask_hint.go

```diff
@@ -0,0 +1,244 @@
+package items
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"unsafe"
+
+	"github.com/go-chi/render"
+	"github.com/jinzhu/gorm"
+	"gopkg.in/jose.v1/crypto"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/logging"
+	"github.com/France-ioi/AlgoreaBackend/app/payloads"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+	"github.com/France-ioi/AlgoreaBackend/app/token"
+)
+
+func (srv *Service) askHint(w http.ResponseWriter, r *http.Request) service.APIError {
+	requestData := AskHintRequest{store: srv.Store, publicKey: srv.TokenConfig.PublicKey}
+
+	var err error
+	if err = render.Bind(r, &requestData); err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+	apiError := service.NoError
+	if apiError = checkAskHintUsers(user, &requestData); apiError != service.NoError {
+		return apiError
+	}
+
+	err = srv.Store.InTransaction(func(store *database.DataStore) error {
+		var hasAccess bool
+		var reason error
+		hasAccess, reason, err = store.Items().CheckSubmissionRights(requestData.TaskToken.Converted.LocalItemID, user)
+		service.MustNotBeError(err)
+
+		if !hasAccess {
+			apiError = service.ErrForbidden(reason)
+			return nil // commit! (CheckSubmissionRights() changes the DB sometimes)
+		}
+
+		userItemStore := store.UserItems()
+		err = userItemStore.CreateIfMissing(user.UserID, requestData.TaskToken.Converted.LocalItemID)
+		service.MustNotBeError(err)
+
+		// Get the previous hints requested JSON data
+		var hintsRequestedParsed []payloads.Anything
+		hintsRequestedParsed, err = queryAndParsePreviouslyRequestedHints(&requestData, store, user, r)
+		if err == gorm.ErrRecordNotFound {
+			apiError = service.ErrNotFound(errors.New("can't find previously requested hints info"))
+			return nil // commit
+		}
+		service.MustNotBeError(err)
+
+		// Add the new requested hint to the list if it's not in the list yet
+		hintsRequestedParsed = addHintToListIfNeeded(hintsRequestedParsed, requestData.HintToken.AskedHint)
+
+		var hintsRequestedNew []byte
+		hintsRequestedNew, err = json.Marshal(hintsRequestedParsed)
+		service.MustNotBeError(err)
+		hintsRequestedNewString := string(hintsRequestedNew)
+		requestData.TaskToken.HintsRequested = &hintsRequestedNewString
+		hintsGivenCountString := strconv.Itoa(len(hintsRequestedParsed))
+		requestData.TaskToken.HintsGivenCount = &hintsGivenCountString
+
+		columnsToUpdate := map[string]interface{}{
+			"sHintsRequested":            hintsRequestedNew,
+			"nbHintsCached":              len(hintsRequestedParsed),
+			"nbTasksWithHelp":            1,
+			"sAncestorsComputationState": "todo",
```

Comment:
Is the hint count taken into account for access? I don't get why it has to be recomputed, but I'm sure there is a reason 🙂

Reply:
ComputeAllUserItems() does this for rows with sAncestorsComputationState="todo":
```
users_items.sHintsRequested = IF(groups_attempts.ID IS NOT NULL,
							groups_attempts.sHintsRequested, users_items.sHintsRequested)
```
So it's not about the hint count, but about sHintsRequested. It can be overwritten with groups_attempts.sHintsRequested's value

Reply:
Looks like a mess. So, if idAttempt is empty in the task token, we generate a hint token, but forget about the given hint...